### ISSUE TITLE
plugin development 

### ISSUE THREAD

**Issue by jbrwn** [mapnik/mapnik#2797]
I am pretty close to submitting a pull requests for a [microsoft sql server plugin](https://github.com/jbrwn/mapnik/tree/mssql/plugins/input/mssql) and would like to get some guidance on contributing.  Looks like the mapnik repo has been rejiggered quite a bit in the last few weeks so i want to make sure I am going down the right path.  Specific questions i have include:

"default" plugins
-
What is the criteria for a default plugin?  I'd like it if my plugin was built by default and included in mapnik-sdk packaging, etc... but i understand that there is probably some discretion involved on what gets included. Is it just a matter of the plugin getting integrated in to the build and packaging process?  

testing
-
Looks like there has been a lot of activity around reorganizing tests lately. i.e, - #2789 and #2790.  Any updates in this area and a description of what kind of tests should be provided for plugins would be appreciated.  

For my mssql plugin functional tests similar to the [python-mapnik postgis tests](https://github.com/mapnik/python-mapnik/blob/master/test/python_tests/postgis_test.py) would have limitations since it requires a local database engine install.   This is especially true for linux/os x builds where the plugin compiles but no database engine can be installed locally.   Would it be enough to only run tests when building on windows?

Seems like another option is to use a more unit test style approach with mock objects/dependency injection/etc... but that seems like it would be a pretty big undertaking.

Thanks in advance for the guidance.

**Comment by flippmoke** [comment]
@springmeyer and I have been talking a lot about all of this recently, I will let him provide all the details once he gets to them, but ideally we want to find a good way to for plugin developers to develop plugins in their own repositories and then simply be able to drop them into a plugins directory. This would be mean that the plugins could be tested seperately from core. 

Part of this means that we would have to maintain a promise of ABI compatibility for the plugins at least on a release basis. The first official release of mapnik 3.x hopefully will include this.

**Comment by jbrwn** [comment]
@flippmoke - thanks for the quick response.  
>ideally we want to find a good way to for plugin developers to develop plugins in their own repositories and then simply be able to drop them into a plugins directory.

Should i interpret this to mean that new plugins will not be distributed as part of the "mapnik-sdk" packages?  My concern here is that users will have to go through a separate install for adding a new plugin to applications that use the mapnik-sdk (mapbox studio, python-mapnik, etc...).  It would be nicer if applications just worked out of the box (i.e., the plugin is included in the mapnik-sdk).

>The first official release of mapnik 3.x hopefully will include this.

So basically just sit tight on plugin development right now?   I'm willing to help out and do whatever needs to be done to streamline this process.   [MSSQL access via OGR is basically broken now in all mapnik products.](https://github.com/mapbox/windows-builds/issues/37) It would be nice to make some progress in this area.

**Comment by flippmoke** [comment]
@jbrwn Well ideally since the plugins, are registered at runtime even in the node and python bindings, we should be able to help people establish packages that simply add those required pieces separately through npm or pip. 

In my little ideal world, I would love to break more things out because it is easier to test individually. We are trying to improve our CI and still make it easy for development as well through this. We are planning on writing a blog post on this on mapnik.org sometime soon, but for now you can see the hello-world plugin that @springmeyer recently updated:

https://github.com/mapnik/hello-world-input-plugin

There are changes that need to be made in the python bindings so any plugin can be simply registered at runtime and work properly, but that is a good start. Ideally we would like to see each plugin have its own tests and CI. 

If you are interested in seeing somewhat how we use our CI stuff, you could look here https://github.com/mapnik/python-mapnik/blob/master/bootstrap.sh at our use of "mason". Currently the latest passing mapnik build on travis is bundled each time mapnik is built as "latest" build version in mason. That might make testing your plugin easier. 

Also I wanted to make sure and ping @jwomeara on this conversation because they are in a similar situation.

**Comment by jbrwn** [comment]
>Well ideally since the plugins, are registered at runtime even in the node and python bindings, we should be able to help people establish packages that simply add those required pieces separately through npm or pip.

Ok so let me back up here for a second.   [My understanding is that mapnik is released as a sdk package that contains binaries and header files.](http://mapnik.s3.amazonaws.com/index.html?path=dist/) This package also contains the binaries for a core set of plugins.  Mapnik-sdk is then used downstream to build out applications and language bindings like python-mapnik and node-mapnik.  The core set of plugins is included in these language binding packages, applications, etc...

Is this approach being reworked?

**Comment by flippmoke** [comment]
I think the ideal way going on in the future will be to have the different input plugin binaries avaiable as distributable packages in node and python. Currently the layout for any mapnik library you download into a `lib` directory is as such:

```
libmapnik.so
mapnik/input/csv.input
mapnik/input/geojson.input
etc...
```

The idea will be to break out some of the plugins that end up in the `input` directory into their own repositories. This allows them to version and configure themselves independent of core. 

The python and node binaries register these at runtime, for example in the python binaries. 
https://github.com/mapnik/python-mapnik/blob/master/mapnik/__init__.py#L1072

What currently works in the node binaries and not the python is if you simply drop a `*.input` file into the bindings `input` directory it will simply just work. (Some modifications are needing in python bindings coming soon).  

If you want to distribute these are their own python or node package it would be simple as well, simply dump x file into a directory and you are done. Its more complex then having it all in core right now, but we are trying to balance:

* Limiting the amount of things in core to make things easier to test
* Limiting the default installs for users
* Simplifying distribution for end package users
* Making it easy to understand how to develop new plugins for mapnik

Among many other things. By the way, I am very glad you are asking these questions :+1: I want to make sure developers like you are able to expand the capabilities of mapnik, but still have it manageable for us core developers! Thanks.

**Comment by jbrwn** [comment]
Thanks @flippmoke.  I'll be keeping an eye on this as the concept evolves.   

>If you want to distribute these are their own python or node package it would be simple as well

Agreed.  But what about other language bindings/packaging systems or the c++ lib for linux/os x/windows?   The list of installer packages needed to support the entire mapnik ecosystem starts to get pretty big.  

I will include the mssql plugin I am developing in the [net-mapnik](https://github.com/jbrwn/NET-Mapnik) bindings I maintain.  I would love to push this plugin back upstream so it is available in other bindings and distributions but i'll be honest: it seems like a lot of work to track other projects and maintain plugin installers for each one.  

Please keep me in the loop on this.  I'd love to be a test case.
